# NB_structural_variants

## Somatic structural variation targets neurodevelopmental genes and identifies SHANK2 as a tumor suppressor in neuroblastoma
Gonzalo Lopez<sup>1,2+</sup>, Karina L. Conkrite<sup>1,3+</sup>, Miriam Doepner<sup>1,3</sup>, Komal S. Rathi<sup>4</sup>, Apexa Modi<sup>1,3,11</sup>, 
Zalman Vaksman<sup>1,3,4</sup>, Lance M. Farra<sup>1,3</sup>, Eric Hyson<sup>1,3</sup>, Moataz Noureddine<sup>1,3</sup>, Jun S. Wei<sup>5</sup>, Malcolm A. Smith<sup>10</sup>, Shahab Asgharzadeh<sup>6,7</sup>, Robert C. Seeger<sup>6,7</sup>, Javed Khan<sup>5</sup>, Jaime Guidry Auvil<sup>9</sup>, Daniela S. Gerhard<sup>9</sup>, John M. Maris<sup>1,3,11,12</sup>, Sharon J. Diskin<sup>1,3,4,10,11*</sup>

1.	Division of Oncology, Children’s Hospital of Philadelphia, Philadelphia, PA, USA.
2.	Department of Genetics and Genomic Sciences and Icahn Institute for Data Science and Genomic Technology, Icahn School of Medicine at Mount Sinai, New York, NY 10029, USA
3.	Center for Childhood Cancer Research, Children’s Hospital of Philadelphia, Philadelphia, PA, USA.
4.	Department of Biomedical and Health Informatics, Children’s Hospital of Philadelphia, Philadelphia, PA, USA.
5.	Oncogenomics Section, Genetics Branch, Center for Cancer Research, National Cancer Institute, Bethesda, MD, USA
6.	Division of Hematology, Oncology and Blood and Marrow Transplantation, Keck School of Medicine of the University of Southern California, Los Angeles, CA, USA.
7.	The Saban Research Institute, Children’s Hospital of Los Angeles, Los Angeles, CA, USA.
8.	Department of Pediatrics, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA, USA.
9.	Office of Cancer Genomics, National Cancer Institute, Bethesda, MD, USA.
10.	Cancer Therapy Evaluation Program, National Cancer Institute, Bethesda, MD, USA.
11.	Genomics and Computational Biology, Biomedical Graduate Studies, Perelman School of Medicine, University of Pennsylvania, Philadelphia, PA, USA.
12.	Abramson Family Cancer Research Institute, Perelman School of Medicine at the University of Pennsylvania, Philadelphia, PA, USA.


This repository contains all code and processed data necessary to reproduce the main and supplementaru figures as well as supplementary tables of the above cited manuscript.

## Download using Git

`git clone https://github.com/diskin-lab-chop/NB_structural_variants`


## R code :
### R/01-nbl_somatic_SV_FUNCTIONS.R
Contain required functions used by the main scripts 02, 03 & 04
### R/02-nbl_somatic_SJBP_analysis.R
Structural variant analysis applied to all [TARGET](https://ocg.cancer.gov/programs/target) WGS cancer datasets; The code implements variant filtering (common variant and artifact removal) and variant annotations (gene-level recurrent alterations) from SJ-BP (sequence junction based variants)
### R/03-nbl_somatic_CNBP_RDBP_analyses.R
Analysis of CNV breakpoints across [TARGET](https://ocg.cancer.gov/programs/target) WGS cancer datasets and SNP array neuroblastoma datasets; The code implements identification of breakpoints and annotation of recurrently altered genes (amplifications, deletions, etc)
### R/04-nbl_somatic_SV_FIGURES.R
Uses the output from SV and CNV analyses to generate additional analyses and figures from manuscript.
###  R/my_stat_functions.r, R/my_survival.R and R/heatmap3.R 
Additional functions for statistics and heatmap plotting

## TARGET processed data:
### allSomatic_cgi_SV_data_V2.rda                                          
Object containing all TARGET tumor variant calls generated by CGI (unfiltered)
### kallisto_nbltpm_refseq_hg19_171.rda
RNA-seq quantification matrix for the TARGET neuroblastoma dataset
### ln_mexp.rda
HumanExon array expression data matrix from the TARGET neuroblastoma dataset
### clinical_COG_20181129_plus_TARGET_20180331_plus_SCA_20180619_pheno.rda 
Clinical info from the neuroblastoma TARGET cohort
### cnv_segmentation_CGI_081318.rda                                        
Segmentation CNV data for the 135 whole genomes (CGI)
### cgiSegDataTumorProcessed_k2_g1000.rda                                  
Segmentation CNV data for TARGET tumors (except neuroblastoma)
### cnv_segmentation_SNP_081318.rda
Segmentation CNV data for the 914 SNP array samples
### SraRunTable_wgs_alltumors.txt & usi_2targetID.txt                                          
Crossreference IDs for TARGET cohort

## Downstream analysis data:
### BP_analysis_Nov15_19.rda
CN-BP and RD-BP analyses derived from segmentation data breakpoints including recurrentluy altered genes
### SV_analysis_Oct31_18.rda                                               
SJ-BP analyses derived from discordantly aligned read pair analysis from CGI including filtered variants and recurrently altered genes
### extended_cgi_Aug_1_18.maf
MAF file containing filtered SNVs and small variants in the 135 CGI neuroblastoma dataset
### FUNCTION                                                               
A folder containing several tables with the pathway analysis performed in ToppGene
### CGI_low_coverage_regions.txt                                           
List of low coverage regions from hg19 discarded in the study

## Additional external data:
### GRCh37_hg19_variants_2016-05-15-small.txt   
The database of genomic variants
### GSE62564_SEQC_NB_RNA-Seq_log2RPM.txt.gz       
SEQC expression data matrix
### seqc_survival_clinical.txt
SEQC clinical info
### autism_disorder_genes.txt     
The list of genes involved in autism spectrum disorder from the ToppGene Disease collection
### cancer_gene_census_v85.csv                                             
The COSMIC cancer gene census
### ucsc_hg19_refseq_genes_exon_df_Oct31_2018.rda
Genes and gene exon annotations
### MSigDB_v6.1_human.rda                                                  
The Broad Institute MSigDB pathway and gene set collection



